File: sdk/android/frak-sdk-ui/src/main/kotlin/id/frak/sdk/ui/SharingSheetSkeleton.kt
package id.frak.sdk.ui

import androidx.compose.animation.core.RepeatMode
import androidx.compose.animation.core.animateFloat
import androidx.compose.animation.core.infiniteRepeatable
import androidx.compose.animation.core.rememberInfiniteTransition
import androidx.compose.animation.core.tween
import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.BottomSheetDefaults
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Surface
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.alpha
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Shape
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp

/**
 * What the sheet shows until the hosted page has painted. Opaque, and stacked over the web view for
 * the whole load: the web view is transparent, so an unpainted one shows nothing at all.
 */
@OptIn(androidx.compose.material3.ExperimentalMaterial3Api::class)
@Composable
internal fun SharingSheetSkeleton(modifier: Modifier = Modifier) {
    val shimmer = rememberInfiniteTransition(label = "frak-sharing-skeleton")
    val alpha by shimmer.animateFloat(
        initialValue = MIN_ALPHA,
        targetValue = MAX_ALPHA,
        animationSpec =
            infiniteRepeatable(
                animation = tween(durationMillis = PULSE_MILLIS),
                repeatMode = RepeatMode.Reverse,
            ),
        label = "frak-sharing-skeleton-alpha",
    )

    Surface(
        modifier = modifier.fillMaxSize(),
        // Compose-drawn, so it can round the corners the rectangular web view cannot, until the
        // page paints and takes over with the matching CSS radius.
        shape = SheetCornerShape,
        color = BottomSheetDefaults.ContainerColor,
    ) {
        Column(
            modifier =
                Modifier
                    .fillMaxSize()
                    // Top inset clears the grab strip the sheet draws over this.
                    .padding(start = 20.dp, end = 20.dp, top = 36.dp, bottom = 20.dp)
                    .alpha(alpha),
            verticalArrangement = Arrangement.spacedBy(16.dp),
        ) {
            // Merchant row: logo, name.
            Row(horizontalArrangement = Arrangement.spacedBy(12.dp)) {
                Block(width = Modifier.size(28.dp), height = 28.dp, shape = CircleShape)
                Block(width = Modifier.weight(0.5f), height = 28.dp)
            }

            // The reward headline, the page's largest element.
            Block(width = Modifier.fillMaxWidth(), height = 92.dp, shape = RoundedCornerShape(16.dp))

            // Product cards.
            Row(
                modifier = Modifier.fillMaxWidth(),
                horizontalArrangement = Arrangement.spacedBy(12.dp),
            ) {
                Block(width = Modifier.weight(1f), height = 108.dp, shape = RoundedCornerShape(12.dp))
                Block(width = Modifier.weight(1f), height = 108.dp, shape = RoundedCornerShape(12.dp))
            }

            // The three-step explainer.
            repeat(STEP_COUNT) {
                Row(horizontalArrangement = Arrangement.spacedBy(12.dp)) {
                    Block(width = Modifier.size(20.dp), height = 20.dp, shape = CircleShape)
                    Block(width = Modifier.weight(1f), height = 20.dp)
                }
            }

            Spacer(Modifier.weight(1f))

            // Share / Copy footer.
            Block(width = Modifier.fillMaxWidth(), height = 48.dp, shape = RoundedCornerShape(24.dp))
        }
    }
}

@Composable
private fun Block(
    width: Modifier,
    height: Dp,
    shape: Shape = RoundedCornerShape(8.dp),
) {
    Spacer(
        modifier =
            width
                .height(height)
                .clip(shape)
                .background(MaterialTheme.colorScheme.surfaceVariant),
    )
}

private const val MIN_ALPHA = 0.45f
private const val MAX_ALPHA = 0.85f
private const val PULSE_MILLIS = 700
private const val STEP_COUNT = 3

File: sdk/ios/Sources/FrakSDKUI/SharingSheetSkeleton.swift
#if canImport(UIKit)
    import SwiftUI

    /// What the sheet shows until the hosted page has actually painted.
    ///
    /// Opaque and stacked over the web view for the whole load, because a `WKWebView` paints its
    /// background before it has content. Mirrors `SharingSheetSkeleton.kt`.
    struct SharingSheetSkeleton: View {
        @State private var pulsing = false

        var body: some View {
            ZStack {
                // Opaque: this covers a web view rather than floating over the merchant's screen.
                Color(.systemBackground)

                VStack(alignment: .leading, spacing: 16) {
                    // Merchant row.
                    HStack(spacing: 12) {
                        Block(width: 28, height: 28, cornerRadius: 14)
                        Block(height: 28)
                            .frame(maxWidth: 160)
                        Spacer(minLength: 0)
                    }

                    // The reward headline, the page's largest element.
                    Block(height: 92, cornerRadius: 16)

                    // Product cards.
                    HStack(spacing: 12) {
                        Block(height: 108, cornerRadius: 12)
                        Block(height: 108, cornerRadius: 12)
                    }

                    // The three-step explainer.
                    ForEach(0..<Self.stepCount, id: \.self) { _ in
                        HStack(spacing: 12) {
                            Block(width: 20, height: 20, cornerRadius: 10)
                            Block(height: 20)
                        }
                    }

                    Spacer(minLength: 0)

                    // Share / Copy footer.
                    Block(height: 48, cornerRadius: 24)
                }
                // Top inset clears the sheet's own grabber, which is drawn over this.
                .padding(EdgeInsets(top: 36, leading: 20, bottom: 20, trailing: 20))
                .opacity(pulsing ? Self.maxOpacity : Self.minOpacity)
                .animation(
                    .easeInOut(duration: Self.pulseDuration).repeatForever(autoreverses: true),
                    value: pulsing
                )
            }
            // Started from `onAppear`: an animation on a state that never changes never runs.
            .onAppear { pulsing = true }
            .accessibilityHidden(true)
        }

        private static let minOpacity = 0.45
        private static let maxOpacity = 0.85
        private static let pulseDuration = 0.7
        private static let stepCount = 3
    }

    /// One grey bar. `width` nil means "take what the row gives me".
    private struct Block: View {
        var width: CGFloat?
        let height: CGFloat
        var cornerRadius: CGFloat = 8

        var body: some View {
            RoundedRectangle(cornerRadius: cornerRadius, style: .continuous)
                .fill(Color(.secondarySystemBackground))
                .frame(width: width, height: height)
        }
    }
#endif
